core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -0,0 +1,227 @@
+    checkAndIncrementLong(toRecord, fromRecord, Column.POSITION_DELETE_RECORD_COUNT);
+    checkAndIncrementInt(toRecord, fromRecord, Column.POSITION_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toRecord, fromRecord, Column.EQUALITY_DELETE_RECORD_COUNT);
+    checkAndIncrementInt(toRecord, fromRecord, Column.EQUALITY_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toRecord, fromRecord, Column.TOTAL_RECORD_COUNT);
+    if (fromRecord.get(Column.LAST_UPDATED_AT.ordinal()) != null) {
+      if (toRecord.get(Column.LAST_UPDATED_AT.ordinal()) == null
+          || ((long) toRecord.get(Column.LAST_UPDATED_AT.ordinal())
+              < (long) fromRecord.get(Column.LAST_UPDATED_AT.ordinal()))) {
+        toRecord.set(
+            Column.LAST_UPDATED_AT.ordinal(), fromRecord.get(Column.LAST_UPDATED_AT.ordinal()));
+        toRecord.set(
+            Column.LAST_UPDATED_SNAPSHOT_ID.ordinal(),
+            fromRecord.get(Column.LAST_UPDATED_SNAPSHOT_ID.ordinal()));
+      }
+    }
+  }
+
+  /**
+   * Converts the given {@link PartitionData} into a {@link Record} based on the specified partition
+   * schema.
+   *
+   * @param partitionSchema the schema defining the structure of the partition data.
+   * @param partitionData the data to be converted into a Record.
+   * @return a Record that represents the partition data as per the given schema.
+   */
+  public static Record partitionDataToRecord(
+      Types.StructType partitionSchema, PartitionData partitionData) {
+    GenericRecord genericRecord = GenericRecord.create(partitionSchema);
+    for (int index = 0; index < partitionData.size(); index++) {
+      genericRecord.set(index, partitionData.get(index));
+    }
+
+    return genericRecord;
+  }
+
+  private static Record fromManifestEntry(
+      ManifestEntry<?> entry, Table table, Schema recordSchema) {
+    GenericRecord record = GenericRecord.create(recordSchema);
+    Types.StructType partitionType =
+        recordSchema.findField(Column.PARTITION.name()).type().asStructType();
+    PartitionData partitionData = coercedPartitionData(entry.file(), table.specs(), partitionType);
+    record.set(Column.PARTITION.ordinal(), partitionDataToRecord(partitionType, partitionData));
+    record.set(Column.SPEC_ID.ordinal(), entry.file().specId());
+
+    Snapshot snapshot = table.snapshot(entry.snapshotId());
+    if (snapshot != null) {
+      record.set(Column.LAST_UPDATED_SNAPSHOT_ID.ordinal(), snapshot.snapshotId());
+      record.set(Column.LAST_UPDATED_AT.ordinal(), snapshot.timestampMillis());
+    }
+
+    switch (entry.file().content()) {
+      case DATA:
+        record.set(Column.DATA_FILE_COUNT.ordinal(), 1);
+        record.set(Column.DATA_RECORD_COUNT.ordinal(), entry.file().recordCount());
+        record.set(Column.TOTAL_DATA_FILE_SIZE_IN_BYTES.ordinal(), entry.file().fileSizeInBytes());
+        break;
+      case POSITION_DELETES:
+        record.set(Column.POSITION_DELETE_FILE_COUNT.ordinal(), 1);
```

Comment:
If there are no deletes, what kind of values will we have as delete counts? Null or 0?

Reply:
Null, as mentioned in https://github.com/apache/iceberg/pull/10176#discussion_r1698066417, 

also added the tests.